File: src/NerdBank.GitVersioning/SemanticVersion.cs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Diagnostics;
using System.Text.RegularExpressions;
using Validation;

namespace Nerdbank.GitVersioning;

/// <summary>
/// Describes a version with an optional unstable tag.
/// </summary>
[DebuggerDisplay("{DebuggerDisplay,nq}")]
public class SemanticVersion : IEquatable<SemanticVersion>
{
    /// <summary>
    /// The regular expression with capture groups for semantic versioning.
    /// It considers PATCH to be optional and permits the 4th Revision component.
    /// </summary>
    /// <remarks>
    /// Parts of this regex inspired by <see href="https://github.com/sindresorhus/semver-regex/blob/master/index.js">this code</see>.
    /// </remarks>
    private static readonly Regex FullSemVerPattern = new Regex(@"^v?(?<major>0|[1-9][0-9]*)\.(?<minor>0|[1-9][0-9]*)(?:\.(?<patch>0|[1-9][0-9]*)(?:\.(?<revision>0|[1-9][0-9]*))?)?(?<prerelease>-[\da-z\-]+(?:\.[\da-z\-]+)*)?(?<buildMetadata>\+[\da-z\-]+(?:\.[\da-z\-]+)*)?$", RegexOptions.IgnoreCase);

    /// <summary>
    /// The regex pattern that a prerelease must match.
    /// </summary>
    /// <remarks>
    /// Keep in sync with the regex for the version field found in the version.schema.json file.
    /// </remarks>
    private static readonly Regex PrereleasePattern = new Regex("-(?:[\\da-z\\-]+|\\{height\\})(?:\\.(?:[\\da-z\\-]+|\\{height\\}))*", RegexOptions.IgnoreCase);

    /// <summary>
    /// The regex pattern that build metadata must match.
    /// </summary>
    /// <remarks>
    /// Keep in sync with the regex for the version field found in the version.schema.json file.
    /// </remarks>
    private static readonly Regex BuildMetadataPattern = new Regex("\\+(?:[\\da-z\\-]+|\\{height\\})(?:\\.(?:[\\da-z\\-]+|\\{height\\}))*", RegexOptions.IgnoreCase);

    /// <summary>
    /// The regular expression with capture groups for semantic versioning,
    /// allowing for macros such as {height}.
    /// </summary>
    /// <remarks>
    /// Keep in sync with the regex for the version field found in the version.schema.json file.
    /// </remarks>
    private static readonly Regex FullSemVerWithMacrosPattern = new Regex("^v?(?<major>0|[1-9][0-9]*)\\.(?<minor>0|[1-9][0-9]*)(?:\\.(?<patch>0|[1-9][0-9]*)(?:\\.(?<revision>0|[1-9][0-9]*))?)?(?<prerelease>" + PrereleasePattern + ")?(?<buildMetadata>" + BuildMetadataPattern + ")?$", RegexOptions.IgnoreCase);

    /// <summary>
    /// Initializes a new instance of the <see cref="SemanticVersion"/> class.
    /// </summary>
    /// <param name="version">The numeric version.</param>
    /// <param name="prerelease">The prerelease, with leading - character.</param>
    /// <param name="buildMetadata">The build metadata, with leading + character.</param>
    public SemanticVersion(Version version, string prerelease = null, string buildMetadata = null)
    {
        Requires.NotNull(version, nameof(version));
        VerifyPatternMatch(prerelease, PrereleasePattern, nameof(prerelease));
        VerifyPatternMatch(buildMetadata, BuildMetadataPattern, nameof(buildMetadata));

        this.Version = version;
        this.Prerelease = prerelease ?? string.Empty;
        this.BuildMetadata = buildMetadata ?? string.Empty;
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="SemanticVersion"/> class.
    /// </summary>
    /// <param name="version">The x.y.z numeric version.</param>
    /// <param name="prerelease">The prerelease, with leading - character.</param>
    /// <param name="buildMetadata">The build metadata, with leading + character.</param>
    public SemanticVersion(string version, string prerelease = null, string buildMetadata = null)
        : this(new Version(version), prerelease, buildMetadata)
    {
    }

    /// <summary>
    /// Identifies the various positions in a semantic version.
    /// </summary>
    public enum Position
    {
        /// <summary>
        /// The <see cref="Version.Major"/> component.
        /// </summary>
        Major,

        /// <summary>
        /// The <see cref="Version.Minor"/> component.
        /// </summary>
        Minor,

        /// <summary>
        /// The <see cref="Version.Build"/> component.
        /// </summary>
        Build,

        /// <summary>
        /// The <see cref="Version.Revision"/> component.
        /// </summary>
        Revision,

        /// <summary>
        /// The <see cref="Prerelease"/> portion of the version.
        /// </summary>
        Prerelease,

        /// <summary>
        /// The <see cref="BuildMetadata"/> portion of the version.
        /// </summary>
        BuildMetadata,
    }

    /// <summary>
    /// Gets the version.
    /// </summary>
    public Version Version { get; }

    /// <summary>
    /// Gets an unstable tag (with the leading hyphen), if applicable.
    /// </summary>
    /// <value>A string with a leading hyphen or the empty string.</value>
    public string Prerelease { get; }

    /// <summary>
    /// Gets the build metadata (with the leading plus), if applicable.
    /// </summary>
    /// <value>A string with a leading plus or the empty string.</value>
    public string BuildMetadata { get; }

    /// <summary>
    /// Gets the position in a computed version that the version height should appear.
    /// </summary>
    public SemanticVersion.Position? VersionHeightPosition
    {
        get
        {
            if (this.Prerelease?.Contains(VersionOptions.VersionHeightPlaceholder) ?? false)
            {
                return SemanticVersion.Position.Prerelease;
            }
            else if (this.Version.Build == -1)
            {
                return SemanticVersion.Position.Build;
            }
            else if (this.Version.Revision == -1)
            {
                return SemanticVersion.Position.Revision;
            }
            else
            {
                return null;
            }
        }
    }

    /// <summary>
    /// Gets the position in a computed version that the first 16 bits of a git commit ID should appear, if any.
    /// </summary>
    internal SemanticVersion.Position? GitCommitIdPosition
    {
        get
        {
            // We can only store the git commit ID info after there was a place to put the version height.
            // We don't want to store the commit ID (which is effectively a random integer) in the revision slot
            // if the version height does not appear, or only appears later (in the -prerelease tag) since that
            // would mess up version ordering.
            if (this.VersionHeightPosition == SemanticVersion.Position.Build)
            {
                return SemanticVersion.Position.Revision;
            }
            else
            {
                return null;
            }
        }
    }

    /// <summary>
    /// Gets a value indicating whether this instance is the default "0.0" instance.
    /// </summary>
    internal bool IsDefault => this.Version?.Major == 0 && this.Version.Minor == 0 && this.Version.Build == -1 && this.Version.Revision == -1 && this.Prerelease is null && this.BuildMetadata is null;

    /// <summary>
    /// Gets the debugger display for this instance.
    /// </summary>
    private string DebuggerDisplay => this.ToString();

    /// <summary>
    /// Parses a semantic version from the given string.
    /// </summary>
    /// <param name="semanticVersion">The value which must wholly constitute a semantic version to succeed.</param>
    /// <param name="version">Receives the semantic version, if found.</param>
    /// <returns><see langword="true"/> if a semantic version is found; <see langword="false"/> otherwise.</returns>
    public static bool TryParse(string semanticVersion, out SemanticVersion version)
    {
        Requires.NotNullOrEmpty(semanticVersion, nameof(semanticVersion));

        Match m = FullSemVerWithMacrosPattern.Match(semanticVersion);
        if (m.Success)
        {
            int major = int.Parse(m.Groups["major"].Value);
            int minor = int.Parse(m.Groups["minor"].Value);
            string patch = m.Groups["patch"].Value;
            string revision = m.Groups["revision"].Value;
            Version systemVersion = patch.Length > 0
                ? revision.Length > 0 ? new Version(major, minor, int.Parse(patch), int.Parse(revision)) : new Version(major, minor, int.Parse(patch))
                : new Version(major, minor);
            string prerelease = m.Groups["prerelease"].Value;
            string buildMetadata = m.Groups["buildMetadata"].Value;
            version = new SemanticVersion(systemVersion, prerelease, buildMetadata);
            return true;
        }

        version = null;
        return false;
    }

    /// <summary>
    /// Parses a semantic version from the given string.
    /// </summary>
    /// <param name="semanticVersion">The value which must wholly constitute a semantic version to succeed.</param>
    /// <returns>An instance of <see cref="SemanticVersion"/>, initialized to the value specified in <paramref name="semanticVersion"/>.</returns>
    public static SemanticVersion Parse(string semanticVersion)
    {
        SemanticVersion result;
        Requires.Argument(TryParse(semanticVersion, out result), nameof(semanticVersion), "Unrecognized or unsupported semantic version.");
        return result;
    }

    /// <summary>
    /// Checks equality against another object.
    /// </summary>
    /// <param name="obj">The other instance.</param>
    /// <returns><see langword="true"/> if the instances have equal values; <see langword="false"/> otherwise.</returns>
    public override bool Equals(object obj)
    {
        return this.Equals(obj as SemanticVersion);
    }

    /// <summary>
    /// Gets a hash code for this instance.
    /// </summary>
    /// <returns>The hash code.</returns>
    public override int GetHashCode()
    {
        return this.Version.GetHashCode() + this.Prerelease.GetHashCode();
    }

    /// <summary>
    /// Prints this instance as a string.
    /// </summary>
    /// <returns>A string representation of this object.</returns>
    public override string ToString()
    {
        return this.Version + this.Prerelease + this.BuildMetadata;
    }

    /// <summary>
    /// Checks equality against another instance of this class.
    /// </summary>
    /// <param name="other">The other instance.</param>
    /// <returns><see langword="true"/> if the instances have equal values; <see langword="false"/> otherwise.</returns>
    public bool Equals(SemanticVersion other)
    {
        if (other is null)
        {
            return false;
        }

        return this.Version == other.Version
            && this.Prerelease == other.Prerelease
            && this.BuildMetadata == other.BuildMetadata;
    }

    /// <summary>
    /// Tests whether two <see cref="SemanticVersion" /> instances are compatible enough that version height is not reset
    /// when progressing from one to the next.
    /// </summary>
    /// <param name="first">The first semantic version.</param>
    /// <param name="second">The second semantic version.</param>
    /// <param name="versionHeightPosition">The position within the version where height is tracked.</param>
    /// <returns><see langword="true"/> if transitioning from one version to the next should reset the version height; <see langword="false"/> otherwise.</returns>
    internal static bool WillVersionChangeResetVersionHeight(SemanticVersion first, SemanticVersion second, SemanticVersion.Position versionHeightPosition)
    {
        Requires.NotNull(first, nameof(first));
        Requires.NotNull(second, nameof(second));

        if (first == second)
        {
            return false;
        }

        if (versionHeightPosition == SemanticVersion.Position.Prerelease)
        {
            // The entire version spec must match exactly.
            return !first.Equals(second);
        }

        for (SemanticVersion.Position position = SemanticVersion.Position.Major; position <= versionHeightPosition; position++)
        {
            int expectedValue = ReadVersionPosition(second.Version, position);
            int actualValue = ReadVersionPosition(first.Version, position);
            if (expectedValue != actualValue)
            {
                return true;
            }
        }

        return false;
    }

    internal static int ReadVersionPosition(Version version, Position position)
    {
        Requires.NotNull(version, nameof(version));

        return position switch
        {
            Position.Major => version.Major,
            Position.Minor => version.Minor,
            Position.Build => version.Build,
            Position.Revision => version.Revision,
            _ => throw new ArgumentOutOfRangeException(nameof(position), position, "Must be one of the 4 integer parts."),
        };
    }

    internal int ReadVersionPosition(Position position) => ReadVersionPosition(this.Version, position);

    /// <summary>
    /// Checks whether a given version may have been produced by this semantic version.
    /// </summary>
    /// <param name="version">The version to test.</param>
    /// <returns><see langword="true"/> if the <paramref name="version"/> is a match; <see langword="false"/> otherwise.</returns>
    internal bool IsMatchingVersion(Version version)
    {
        Position lastPositionToConsider = Position.Revision;
        if (this.VersionHeightPosition <= lastPositionToConsider)
        {
            lastPositionToConsider = this.VersionHeightPosition.Value - 1;
        }

        for (Position i = Position.Major; i <= lastPositionToConsider; i++)
        {
            if (this.ReadVersionPosition(i) != ReadVersionPosition(version, i))
            {
                return false;
            }
        }

        return true;
    }

    /// <summary>
    /// Checks whether a particular version number
    /// belongs to the set of versions represented by this semantic version spec.
    /// </summary>
    /// <param name="version">A version, with major and minor components, and possibly build and/or revision components.</param>
    /// <returns><see langword="true"/> if <paramref name="version"/> may have been produced by this semantic version; <see langword="false"/> otherwise.</returns>
    internal bool Contains(Version version)
    {
        return
            this.Version.Major == version.Major &&
            this.Version.Minor == version.Minor &&
            (this.Version.Build == -1 || this.Version.Build == version.Build) &&
            (this.Version.Revision == -1 || this.Version.Revision == version.Revision);
    }

    /// <summary>
    /// Verifies that the prerelease tag follows semver rules.
    /// </summary>
    /// <param name="input">The input string to test.</param>
    /// <param name="pattern">The regex that the string must conform to.</param>
    /// <param name="parameterName">The name of the parameter supplying the <paramref name="input"/>.</param>
    /// <exception cref="ArgumentException">
    /// Thrown if the <paramref name="input"/> does not match the required <paramref name="pattern"/>.
    /// </exception>
    private static void VerifyPatternMatch(string input, Regex pattern, string parameterName)
    {
        Requires.NotNull(pattern, nameof(pattern));

        if (string.IsNullOrEmpty(input))
        {
            return;
        }

        Requires.Argument(pattern.IsMatch(input), parameterName, $"The prerelease must match the pattern \"{pattern}\".");
    }
}

File: src/nerdbank-gitversioning-rs/src/semantic_version.rs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

use std::fmt;
use std::str::FromStr;
use std::sync::LazyLock;

use regex::Regex;

use crate::{Error, ReleaseVersionIncrement, Result, Version};

static SEMANTIC_VERSION_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"(?ix)^v?(?<major>0|[1-9][0-9]*)\.(?<minor>0|[1-9][0-9]*)(?:\.(?<patch>0|[1-9][0-9]*)(?:\.(?<revision>0|[1-9][0-9]*))?)?(?<prerelease>-(?:[0-9a-z-]+|\{height\})(?:\.(?:[0-9a-z-]+|\{height\}))*)?(?<build_metadata>\+(?:[0-9a-z-]+|\{height\})(?:\.(?:[0-9a-z-]+|\{height\}))*)?$",
    )
    .expect("the semantic version pattern is valid")
});

static PRERELEASE_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?i)^-(?:[0-9a-z-]+|\{height\})(?:\.(?:[0-9a-z-]+|\{height\}))*$")
        .expect("the prerelease pattern is valid")
});

static BUILD_METADATA_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?i)^\+(?:[0-9a-z-]+|\{height\})(?:\.(?:[0-9a-z-]+|\{height\}))*$")
        .expect("the build metadata pattern is valid")
});

/// Identifies positions in a semantic version.
#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
#[repr(u8)]
pub enum VersionPosition {
    /// The major numeric component.
    Major,
    /// The minor numeric component.
    Minor,
    /// The build numeric component.
    Build,
    /// The revision numeric component.
    Revision,
    /// The prerelease portion.
    Prerelease,
    /// The build metadata portion.
    BuildMetadata,
}

/// Describes a numeric version with optional prerelease and build metadata.
///
/// Unlike crates that accept only SemVer 2, this type intentionally supports
/// the two- to four-part version grammar used by Nerdbank.GitVersioning, along
/// with the `{height}` macro in prerelease and build metadata identifiers.
#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub struct SemanticVersion {
    /// The numeric portion of the version.
    pub version: Version,
    /// The prerelease portion, including its leading hyphen, or an empty string.
    pub prerelease: String,
    /// The build metadata, including its leading plus, or an empty string.
    pub build_metadata: String,
}

impl SemanticVersion {
    /// Initializes a semantic version, validating the prerelease and build
    /// metadata strings.
    pub fn new(
        version: Version,
        prerelease: impl Into<String>,
        build_metadata: impl Into<String>,
    ) -> Result<Self> {
        let prerelease = prerelease.into();
        let build_metadata = build_metadata.into();
        if !prerelease.is_empty() && !PRERELEASE_PATTERN.is_match(&prerelease) {
            return Err(Error::InvalidFormat(format!(
                "The prerelease '{prerelease}' is invalid."
            )));
        }
        if !build_metadata.is_empty() && !BUILD_METADATA_PATTERN.is_match(&build_metadata) {
            return Err(Error::InvalidFormat(format!(
                "The build metadata '{build_metadata}' is invalid."
            )));
        }

        Ok(Self {
            version,
            prerelease,
            build_metadata,
        })
    }

    /// Parses a semantic version, returning `None` when the input is invalid.
    pub fn try_parse(value: &str) -> Option<Self> {
        value.parse().ok()
    }

    /// Gets the position where version height should appear.
    pub fn version_height_position(&self) -> Option<VersionPosition> {
        if self.prerelease.contains("{height}") {
            Some(VersionPosition::Prerelease)
        } else if self.version.build.is_none() {
            Some(VersionPosition::Build)
        } else if self.version.revision.is_none() {
            Some(VersionPosition::Revision)
        } else {
            None
        }
    }

    /// Gets the position where the first 16 bits of a Git commit ID should
    /// appear, if any.
    pub fn git_commit_id_position(&self) -> Option<VersionPosition> {
        (self.version_height_position() == Some(VersionPosition::Build))
            .then_some(VersionPosition::Revision)
    }

    /// Reads one of the four numeric version positions, using `-1` for an
    /// unspecified build or revision.
    pub fn read_version_position(&self, position: VersionPosition) -> Result<i64> {
        Self::read_position(self.version, position)
    }

    /// Tests whether changing between two semantic versions resets version
    /// height at the specified position.
    pub fn will_version_change_reset_version_height(
        first: &Self,
        second: &Self,
        version_height_position: VersionPosition,
    ) -> Result<bool> {
        if first == second {
            return Ok(false);
        }
        if version_height_position == VersionPosition::Prerelease {
            return Ok(first != second);
        }
        if version_height_position > VersionPosition::Revision {
            return Err(Error::InvalidOperation(
                "Version height must occupy a numeric or prerelease position.".to_owned(),
            ));
        }

        for position in [
            VersionPosition::Major,
            VersionPosition::Minor,
            VersionPosition::Build,
            VersionPosition::Revision,
        ]
        .into_iter()
        .take(version_height_position as usize + 1)
        {
            if first.read_version_position(position)? != second.read_version_position(position)? {
                return Ok(true);
            }
        }

        Ok(false)
    }

    /// Checks whether a numeric version may have been produced by this
    /// semantic version specification.
    pub fn is_matching_version(&self, version: &Version) -> bool {
        let last_position = match self.version_height_position() {
            Some(position) if position <= VersionPosition::Revision => position as usize,
            _ => 4,
        };
        let positions = [
            VersionPosition::Major,
            VersionPosition::Minor,
            VersionPosition::Build,
            VersionPosition::Revision,
        ];

        positions[..last_position].iter().all(|&position| {
            self.read_version_position(position).ok()
                == Self::read_position(*version, position).ok()
        })
    }

    /// Checks whether a numeric version belongs to the set represented by this
    /// semantic version specification.
    pub fn contains(&self, version: &Version) -> bool {
        self.version.major == version.major
            && self.version.minor == version.minor
            && self
                .version
                .build
                .is_none_or(|build| version.build == Some(build))
            && self
                .version
                .revision
                .is_none_or(|revision| version.revision == Some(revision))
    }

    /// Gets a new semantic version with the specified numeric component
    /// incremented.
    ///
    /// Incrementing build is invalid when the build component is unspecified.
    /// Components after the incremented component are reset to zero while the
    /// original component count, prerelease, and build metadata are preserved.
    pub fn increment(&self, increment: ReleaseVersionIncrement) -> Result<Self> {
        if increment == ReleaseVersionIncrement::Build && self.version.build.is_none() {
            return Err(Error::InvalidOperation(format!(
                "Cannot apply build increment to version '{self}'."
            )));
        }

        let mut major = self.version.major;
        let mut minor = self.version.minor;
        let mut build = self.version.build.unwrap_or(0);
        match increment {
            ReleaseVersionIncrement::Major => {
                major = major.checked_add(1).ok_or_else(|| {
                    Error::InvalidOperation("The major version cannot be incremented.".to_owned())
                })?;
                minor = 0;
                build = 0;
            }
            ReleaseVersionIncrement::Minor => {
                minor = minor.checked_add(1).ok_or_else(|| {
                    Error::InvalidOperation("The minor version cannot be incremented.".to_owned())
                })?;
                build = 0;
            }
            ReleaseVersionIncrement::Build => {
                build = build.checked_add(1).ok_or_else(|| {
                    Error::InvalidOperation("The build version cannot be incremented.".to_owned())
                })?;
            }
        }

        let version = if self.version.build.is_none() {
            Version::new(major, minor)
        } else if self.version.revision.unwrap_or(0) > 0 {
            Version::new_with_revision(major, minor, build, 0)
        } else {
            Version::new_with_build(major, minor, build)
        };
        Self::new(
            version,
            self.prerelease.clone(),
            self.build_metadata.clone(),
        )
    }

    /// Sets the first prerelease identifier. The leading hyphen may be
    /// specified or omitted. An empty value removes the first identifier.
    pub fn set_first_prerelease_tag(&self, new_first_tag: &str) -> Result<Self> {
        let suffix = self
            .prerelease
            .find('.')
            .map_or("", |index| &self.prerelease[index..]);
        let mut prerelease = if self.prerelease.is_empty() {
            new_first_tag.to_owned()
        } else {
            format!("{new_first_tag}{suffix}")
        };
        if !prerelease.is_empty() && !prerelease.starts_with('-') {
            prerelease.insert(0, '-');
        }
        Self::new(self.version, prerelease, self.build_metadata.clone())
    }

    /// Returns a copy without any prerelease identifiers.
    pub fn without_prerelease_tags(&self) -> Self {
        Self {
            version: self.version,
            prerelease: String::new(),
            build_metadata: self.build_metadata.clone(),
        }
    }

    /// Converts a SemVer 2 prerelease such as `-beta.5` to a SemVer 1
    /// compatible value such as `-beta-0005`.
    ///
    /// Numeric identifiers are left-padded to at least `padding_size` digits,
    /// and dots are replaced by hyphens.
    pub fn make_prerelease_semver1_compliant(prerelease: &str, padding_size: usize) -> String {
        if prerelease.is_empty() {
            return String::new();
        }

        let body = prerelease.strip_prefix('-').unwrap_or(prerelease);
        let identifiers = body.split('.').map(|identifier| {
            if identifier.bytes().all(|byte| byte.is_ascii_digit()) {
                let normalized = identifier.trim_start_matches('0');
                let normalized = if normalized.is_empty() {
                    "0"
                } else {
                    normalized
                };
                format!("{normalized:0>padding_size$}")
            } else {
                identifier.to_owned()
            }
        });
        format!("-{}", identifiers.collect::<Vec<_>>().join("-"))
    }

    fn read_position(version: Version, position: VersionPosition) -> Result<i64> {
        match position {
            VersionPosition::Major => Ok(version.major as i64),
            VersionPosition::Minor => Ok(version.minor as i64),
            VersionPosition::Build => Ok(version.build_or_unspecified()),
            VersionPosition::Revision => Ok(version.revision_or_unspecified()),
            _ => Err(Error::InvalidOperation(
                "The position must be one of the four numeric parts.".to_owned(),
            )),
        }
    }
}

impl fmt::Display for SemanticVersion {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            formatter,
            "{}{}{}",
            self.version, self.prerelease, self.build_metadata
        )
    }
}

impl FromStr for SemanticVersion {
    type Err = Error;

    fn from_str(value: &str) -> Result<Self> {
        let captures = SEMANTIC_VERSION_PATTERN.captures(value).ok_or_else(|| {
            Error::InvalidFormat(format!(
                "Unrecognized or unsupported semantic version '{value}'."
            ))
        })?;
        let parse_number = |name: &str| {
            captures[name]
                .parse::<i32>()
                .map(|value| value as u32)
                .map_err(|_| {
                    Error::InvalidFormat(format!(
                        "'{value}' contains an invalid version component."
                    ))
                })
        };
        let major = parse_number("major")?;
        let minor = parse_number("minor")?;
        let version = match (captures.name("patch"), captures.name("revision")) {
            (None, None) => Version::new(major, minor),
            (Some(patch), None) => Version::new_with_build(
                major,
                minor,
                patch
                    .as_str()
                    .parse::<i32>()
                    .map(|value| value as u32)
                    .map_err(|_| {
                        Error::InvalidFormat(format!(
                            "'{value}' contains an invalid version component."
                        ))
                    })?,
            ),
            (Some(patch), Some(revision)) => Version::new_with_revision(
                major,
                minor,
                patch
                    .as_str()
                    .parse::<i32>()
                    .map(|value| value as u32)
                    .map_err(|_| {
                        Error::InvalidFormat(format!(
                            "'{value}' contains an invalid version component."
                        ))
                    })?,
                revision
                    .as_str()
                    .parse::<i32>()
                    .map(|value| value as u32)
                    .map_err(|_| {
                        Error::InvalidFormat(format!(
                            "'{value}' contains an invalid version component."
                        ))
                    })?,
            ),
            (None, Some(_)) => unreachable!("the regex requires patch before revision"),
        };
        Self::new(
            version,
            captures
                .name("prerelease")
                .map_or("", |capture| capture.as_str()),
            captures
                .name("build_metadata")
                .map_or("", |capture| capture.as_str()),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_full_supported_grammar() {
        let parsed: SemanticVersion = "v1.2.3.4-pre.5.{height}+build-metadata.id1"
            .parse()
            .unwrap();
        assert_eq!(Version::new_with_revision(1, 2, 3, 4), parsed.version);
        assert_eq!("-pre.5.{height}", parsed.prerelease);
        assert_eq!("+build-metadata.id1", parsed.build_metadata);
        assert_eq!(
            Some(VersionPosition::Prerelease),
            parsed.version_height_position()
        );
        assert_eq!(
            "1.2.3.4-pre.5.{height}+build-metadata.id1",
            parsed.to_string()
        );
    }

    #[test]
    fn rejects_bad_grammar_and_leading_zeroes() {
        for value in [
            "", "1", "01.2", "1.02", "1.2.03", "1.2-$", "1.2-", "1.2-a.", "1.2+$", "1.2+", "1.2+a.",
        ] {
            assert!(SemanticVersion::try_parse(value).is_none(), "{value}");
        }
    }

    #[test]
    fn constructors_validate_suffixes() {
        assert!(SemanticVersion::new(Version::new(1, 2), "-pre", "+build").is_ok());
        assert!(SemanticVersion::new(Version::new(1, 2), "pre", "").is_err());
        assert!(SemanticVersion::new(Version::new(1, 2), "", "build").is_err());
    }

    #[test]
    fn computes_height_and_commit_positions() {
        let two: SemanticVersion = "1.2".parse().unwrap();
        let three: SemanticVersion = "1.2.3".parse().unwrap();
        let four: SemanticVersion = "1.2.3.4".parse().unwrap();
        assert_eq!(Some(VersionPosition::Build), two.version_height_position());
        assert_eq!(
            Some(VersionPosition::Revision),
            two.git_commit_id_position()
        );
        assert_eq!(
            Some(VersionPosition::Revision),
            three.version_height_position()
        );
        assert_eq!(None, four.version_height_position());
    }

    #[test]
    fn checks_reset_matching_and_containment() {
        let spec: SemanticVersion = "1.2".parse().unwrap();
        assert!(spec.is_matching_version(&Version::new_with_revision(1, 2, 15, 42)));
        assert!(!spec.is_matching_version(&Version::new_with_build(2, 2, 15)));
        assert!(spec.contains(&Version::new_with_revision(1, 2, 15, 42)));

        let first: SemanticVersion = "1.2.3".parse().unwrap();
        let second: SemanticVersion = "1.2.4".parse().unwrap();
        assert!(
            SemanticVersion::will_version_change_reset_version_height(
                &first,
                &second,
                VersionPosition::Revision
            )
            .unwrap()
        );
        assert!(
            SemanticVersion::will_version_change_reset_version_height(
                &first,
                &second,
                VersionPosition::Build
            )
            .unwrap()
        );
    }

    #[test]
    fn edits_and_normalizes_prerelease() {
        let version: SemanticVersion = "1.2-alpha.preview+metadata".parse().unwrap();
        assert_eq!(
            "1.2-beta.preview+metadata",
            version
                .set_first_prerelease_tag("beta")
                .unwrap()
                .to_string()
        );
        assert_eq!(
            "1.2+metadata",
            version.without_prerelease_tags().to_string()
        );
        assert_eq!(
            "-beta-0005-0001-foo5",
            SemanticVersion::make_prerelease_semver1_compliant("-beta.5.00001.foo5", 4)
        );
    }

    #[test]
    fn increments_and_preserves_suffixes_and_precision() {
        let two: SemanticVersion = "1.2-tag+metadata".parse().unwrap();
        assert_eq!(
            "2.0-tag+metadata",
            two.increment(ReleaseVersionIncrement::Major)
                .unwrap()
                .to_string()
        );
        assert_eq!(
            "1.3-tag+metadata",
            two.increment(ReleaseVersionIncrement::Minor)
                .unwrap()
                .to_string()
        );
        assert!(two.increment(ReleaseVersionIncrement::Build).is_err());

        let four: SemanticVersion = "1.2.3.4-tag".parse().unwrap();
        assert_eq!(
            "1.2.4.0-tag",
            four.increment(ReleaseVersionIncrement::Build)
                .unwrap()
                .to_string()
        );
    }

    #[test]
    fn managed_increment_and_first_tag_compatibility_vectors() {
        for (input, increment, expected) in [
            ("1.0", ReleaseVersionIncrement::Minor, "1.1"),
            ("1.1", ReleaseVersionIncrement::Major, "2.0"),
            ("1.2.3", ReleaseVersionIncrement::Minor, "1.3.0"),
            ("1.2.3", ReleaseVersionIncrement::Major, "2.0.0"),
            ("1.2.3", ReleaseVersionIncrement::Build, "1.2.4"),
            ("1.2.3.4", ReleaseVersionIncrement::Minor, "1.3.0.0"),
            ("1.2.3.4", ReleaseVersionIncrement::Major, "2.0.0.0"),
            ("1.2.3.4", ReleaseVersionIncrement::Build, "1.2.4.0"),
            (
                "1.2.3-tag+metadata",
                ReleaseVersionIncrement::Build,
                "1.2.4-tag+metadata",
            ),
        ] {
            let actual = input
                .parse::<SemanticVersion>()
                .unwrap()
                .increment(increment)
                .unwrap();
            assert_eq!(expected, actual.to_string(), "{input}");
        }

        for (input, tag, expected) in [
            ("1.2", "pre", "1.2-pre"),
            ("1.2+build", "-pre", "1.2-pre+build"),
            ("1.2-alpha", "beta", "1.2-beta"),
            ("1.2-alpha.preview", "-beta", "1.2-beta.preview"),
            (
                "1.2-alpha.preview+metadata",
                "beta",
                "1.2-beta.preview+metadata",
            ),
            ("1.2-alpha.{height}", "beta", "1.2-beta.{height}"),
            ("1.2-pre", "", "1.2"),
        ] {
            let actual = input
                .parse::<SemanticVersion>()
                .unwrap()
                .set_first_prerelease_tag(tag)
                .unwrap();
            assert_eq!(expected, actual.to_string(), "{input}");
        }
    }
}
